File: Flood Fill.cc
class Solution {
public:
    int oldc;
    
    void fill(vector<vector<int>>& image, int sr, int sc, int colr, int tr, int tc) {
        if (image[sr][sc] == colr) return;
        
        image[sr][sc] = colr;
        
        if (sc + 1 < tc and image[sr][sc+1] == oldc) {
            fill(image, sr, sc+1, colr, tr, tc);
        }
        
        if (sc - 1 >= 0 and image[sr][sc-1] == oldc) {
            fill(image, sr, sc-1, colr, tr, tc);
        }
        
        if (sr - 1 >= 0 and image[sr-1][sc] == oldc) {
            fill(image, sr-1, sc, colr, tr, tc);
        }
        
        if (sr + 1 < tr and image[sr+1][sc] == oldc) {
            fill(image, sr+1, sc, colr, tr, tc);
        }
        
        return;
    }
    
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        int row = image.size();
        int col = image[0].size();
        oldc = image[sr][sc];
        
        fill(image, sr, sc, newColor, row, col);   
        return image;
    }
};

File: Flood Fill.java
class Solution 
{
    
    public int[][] dir = {
        {0,1},
        {0,-1},
        {1,0},
        {-1,0}
    };
    
    public  void dfs(int[][]  image, int sr, int sc, int newColor,int oldColor)
    {
        if(image[sr][sc] == newColor) return;
        image[sr][sc] = newColor;
        
        int n = image.length;
        int m = image[0].length;
        
        for(int i = 0 ; i<4 ;i++)
        {
            int newR = sr+dir[i][0];
            int newC = sc+dir[i][1];
            
            if(newR>=0 && newC>=0 && newR<n && newC<m && image[newR][newC] == oldColor)
            {
                dfs(image,newR,newC,newColor, oldColor);
            }
        }
        
    }
    
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) 
    {
        dfs(image,sr,sc,newColor,  image[sr][sc]);
        return image;
    }
}
